common/lib/xmodule/xmodule/js/src/problem/edit.coffee

```diff
@@ -189,31 +192,129 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
     else
       return template
 
-# We may wish to add insertHeader. Here is Tom's code.
-# function makeHeader() {
-#  var selection = simpleEditor.getSelection();
-#  var revisedSelection = selection + '\n';
-#  for(var i = 0; i < selection.length; i++) {
-#revisedSelection += '=';
-#  }
-#  simpleEditor.replaceSelection(revisedSelection);
-#}
-#
+
   @markdownToXml: (markdown)->
     toXml = `function (markdown) {
       var xml = markdown,
           i, splits, scriptFlag;
 
+      // fix DOS \r\n line endings to look like \n
+      xml = xml.replace(/\r\n/g, '\n');
+
       // replace headers
       xml = xml.replace(/(^.*?$)(?=\n\=\=+$)/gm, '<h1>$1</h1>');
       xml = xml.replace(/\n^\=\=+$/gm, '');
 
-      // group multiple choice answers
+      // Pull out demand hints,  || a hint ||
+      var demandhints = '';
+      xml = xml.replace(/(^\s*\|\|.*?\|\|\s*$\n?)+/gm, function(match) {  // $\n
+          var options = match.split('\n');
+          for (i = 0; i < options.length; i += 1) {
+              var inner = /\s*\|\|(.*?)\|\|/.exec(options[i]);
+              if (inner) {
+                  demandhints += '  <hint>' + inner[1].trim() + '</hint>\n';
+              }
+           }
+           return '';
+      });
+
+      // replace \n+whitespace within extended hint {{ .. }}, by a space, so the whole
+      // hint sits on one line.
+      // This is the one instance of {{ ... }} matching that permits \n
+      xml = xml.replace(/{{(.|\n)*?}}/gm, function(match) {
+          return match.replace(/\r?\n( |\t)*/g, ' ');
```

Comment:
Must: extra spaces get added when the match has several subsequent newlines. Based on my testing, you can use `/\r?\n(\s)*/g` to prevent this from happening. (note that this comment is a 'must' assuming you don't want these extra spaces in your XML. If that doesn't matter, then treat this as a nit).